### ISSUE TITLE
release version 1.2

### ISSUE THREAD

**Issue by koeppea** [libnet/libnet#46]
Hi Sam,

when do you think are you going to release version 1.2 of libnet?

Cheers

**Comment by LocutusOfBorg** [comment]
hi @sam-github , I see libnet tagged on github, but not available on sf.
https://sourceforge.net/projects/libnet-dev/

https://github.com/sam-github/libnet/releases/tag/libnet-1.2

is it out or not? according to git seems more an rc0 rather than the final release...
can you please delete the tag or update it? thanks!

**Comment by koeppea** [comment]
Is there maybe anything we could help you with?

**Comment by wljjn** [comment]
Hi All, what is the current status of this project? Will version 1.2 finally be out?

**Comment by sgeto** [comment]
any weekend now...

**Comment by koeppea** [comment]
@sam-github, as it seems that you're lacking free time to take care of this project, it may be worth sharing responsibilities. I think I recall that @sgeto is willing to support.

**Comment by sgeto** [comment]
@koeppea well put. I could help or take over if needed.

**Comment by sam-github** [comment]
@sgeto still interested? would you like me to give you commit & release access?

**Comment by sgeto** [comment]
Yes. Would be nice. Thank you.

"We are saddened by a bird's cry, but not for a fish's blood.
Blessed are those with voices."

-Mamoru Oshii

**Comment by troglobit** [comment]
Hi, maintainer of nemesis here, I could also help out if you guys need a hand.

**Comment by sam-github** [comment]
@troglobit Yes, we need a hand, thanks! I'll add you as a maintainer. I think releases should move to github from sourceforge (which was the only place I could find back then, but is a pain to work with).

**Comment by troglobit** [comment]
Cool, I'll start digging around and see what needs to be done. Is there any IRC channel or such, or should we use this issue for collaboration? I'm fine with whatever you guys prefer.

**Comment by sam-github** [comment]
github issues is the way to talk. I've been planning on doing some refactoring. The code has an odd structure that used to be useful to me, where it has libnet and libnet-lua both as subdirs and the top-level has content I added, README and build scripts.

I've been planning on stripping the lua out into a seperate repo, and pushing the utils and scripts down into libnet/, leaving the top-level empty, so I can do a `git mv libnet/* ./; git rm libnet`. That should leave the repo in a more maintainable shape. I'll take a shot at that now.

There are a bunch of open autoconf PRs. They don't all work (some broke build for me), so untangling them and getting a travis/circleci/something build going would be pretty useful, if you've the energy for that.

**Comment by troglobit** [comment]
Using GitHub issues for talk is fine by me.   I've read up on a few of the issues and PRs already, and I agree with you.  Cleaning up the structure is a great first step, regardless of the shape of the PRs.  Please finish what you have in the pipeline and scratch those itches! :)

I've started working on a PR to fix issue #1, but I'll await your `git mv` change, and look at setting up Travis-CI (which I'm most experienced with).

**Comment by troglobit** [comment]
@sam-github Not 100% sure, but I think I need to be member of the libnet *org* on GitHub, as well as the libnet project, to be able to create the libnet project on https://travis-ci.org/ ... if you feel up to it, you can login on Travis and create the project yourself, and leave the YML file magic to me.

**Comment by sam-github** [comment]
I flailed around a couple minutes and didn't see how to add a project, and the tutorial (https://docs.travis-ci.com/user/tutorial/) failed me at step 2 and 3... Its made more confusing probably because I'm a member of so many orgs that use travis that I have 100s of "projects" :-(. I added you to the org.

**Comment by troglobit** [comment]
Thanks, I'll see what I can do with it! :-)

Yeah it's a bit confusing.  Possibly you have to grant the libnet org access to Travis-CI in your GitHub settings, which doesn't take effect immediately. At least that's my experience from previous projects/orgs.

**Comment by sam-github** [comment]
Looks like Travis is already approved: https://github.com/organizations/libnet/settings/oauth_application_policy

**Comment by troglobit** [comment]
Yup, just noticed myself, and it also popped up in my Travis profile. Thanks! :-)

**Comment by sam-github** [comment]
I PRed the rearrangement of the source, PTAL at  #86 , this is about all I have time for ATM.

**Comment by troglobit** [comment]
Awesome, great work!  I'll have a look at it and if it builds for me I'll merge it, OK?

**Comment by sam-github** [comment]
It's possible the debian/ directory should be just removed. When I used libnet at Wurldtech, I needed it so we could package it for our internal code-base, but Wurldtech isn't involved in maintaining this (anymore), and I don't work for them, and I very, very much doubt Debian or Ubuntu use these files, so it should probably be deleted. If anyone ever asked to bring them back, it would be easy.

**Comment by troglobit** [comment]
Yeah, I'll have a look at that too later. (Quite late here in norther Europe) Best would be to have an active DD to take care of that part.

Here's the first Travis build btw --> https://travis-ci.org/libnet/libnet/builds \o/
Only failed on building the man pages, probably something silly.

**Comment by sam-github** [comment]
Maintainer builds need an additional `make -D doc doc`. I pushed.

**Comment by sam-github** [comment]
And it will need doxygen installed, if you know how to arrange that.

**Comment by troglobit** [comment]
Heh, just pushed that ;-)

**Comment by troglobit** [comment]
Aaaand it works, thanks would've taken me a while to figure out your maintainer magic there.

**Comment by sam-github** [comment]
nice. rebased #86 and re-pushed, but maybe PR builds aren't on yet.

**Comment by troglobit** [comment]
I reviewed and hand merged it, must have gotten confused by that. Oh well

**Comment by dohnuts** [comment]
I hope 1.2 can be released soon, with snprintf everywhere and the right interface enumeration code for openbsd !

**Comment by troglobit** [comment]
@dohnuts Yeah me too. The intention is get all outstanding PRs merged or least closed, and then enter the release phase with testing. Automatic testing via CI is one possibility, but if it drags out then manual testing by anyone who can pitch in.

Personally I can test on at least Linux and macOS.  Hopefully we find someone to test Windows/msys/cygwin and maybe you can do some testing on OpenBSD, just to verify the incoming build system changes to cause any regressions?

**Comment by koeppea** [comment]
Me and @sgeto can also test from an Ettercap perspective... waiting long to improve the IPv6 probing which depends on IPv6 option headers.

**Comment by troglobit** [comment]
@koeppea Awesome, that would help a lot!  I'm pretty sure we'll do another -rc drop so we can also verify unpack+build of the release tarballs.

**Comment by sam-github** [comment]
From my perspective, I'm OK with just releasing, rather than a rc. The old system was a release candidate series, and I emulated that, but I'm not wedded to it.

If the release has issues, it can be fixed and rereleased, particularly if the process is easy.

Note that sourceforge was the only game in town when I looked years ago, but I'm pretty sure github will host tarballs now, and would be easier to use for publishing, I've just never had time to try.

**Comment by sam-github** [comment]
But I'm also OK with doing a -rc series... :-). I just want it clear that the process should be clear and reproducible, but it doesn't have to be identical to the past if @troglobit you want to adjust it.

**Comment by koeppea** [comment]
Yeah releasing using GitHub is pretty easy and comfortable.

**Comment by troglobit** [comment]
@sam-github Got it, thanks! :)

**Comment by troglobit** [comment]
Btw, @sam-github there seems to already be a tag for 1.2 -->https://github.com/libnet/libnet/releases/tag/libnet-1.2

Should we skip ahead to v1.3 directly, or should we remove the tag?  Personally I never remove tags since, when pushed, they may be used/referenced by other projects.

Also, the tag scheme is `libnet-1.2` rather than just `1.2` or `v1.2`.  Would you be OK with my adopting the latter?

**Comment by dohnuts** [comment]
This is the list of software in openBSD  with libnet-1.1

/usr/ports/net/arping
/usr/ports/net/ettercap
/usr/ports/net/packit
/usr/ports/net/tcptraceroute
/usr/ports/net/yersinia
/usr/ports/security/ipguard
/usr/ports/security/isic

I tested arping and tcptraceroute with 1.2, i wanted to reply yesterday but my hard drive is dead (with my 1.2 ports in it) , doesn't matter much it s somewhere else.

because libnet 1.0 was uncompatible, it will be easy to release 1.2 and let other maintainer choose to upgrade to 1.2, i should be able to ship arping with 1.2.

libnet-1.2 and 1.2 are fine.

**Comment by sam-github** [comment]
I'm happy with `v1.2`

**Comment by troglobit** [comment]
@sam-github I'm sorry to ask again, what should we do about the existing `libnet-1.2` tag?  It is not set on any commit on master, or any other published branch.  Would you be OK if I remove it?

**Comment by sam-github** [comment]
@troglobit Yes, its OK  to remove it.

**Comment by troglobit** [comment]
OK guys, dunno if you've noticed, but things are starting to shape up now.  There's still some work to be done to clean up the docs at https://codedocs.xyz/libnet/libnet/, pull requests welcome as usual, and we're closing up on a release.  I'll still do another -rc, mostly for my own sake, but I'll announce it here when it goes live.

**Comment by dohnuts** [comment]
So arping is using libnet 1.2 fine on openbsd Current

Strangely the build does not create the libnet-config binaries
Historic behavior (bsd renamed with 1.1) :

```
# libnet-config-1.1 --defines
-DLIBNET_BSDISH_OS
```

Notes:

```
[0]-[current]-[/usr/ports/net/arping]
# arping 172.16.1.1
ARPING 172.16.1.1
60 bytes from ac:64:dd:b0:00:03 (172.16.1.1): index=0 time=324.658 usec
60 bytes from ac:64:dd:b0:00:03 (172.16.1.1): index=1 time=329.084 usec
60 bytes from ac:64:dd:b0:00:03 (172.16.1.1): index=2 time=202.764 usec
60 bytes from ac:64:dd:b0:00:03 (172.16.1.1): index=3 time=843.567 usec
60 bytes from ac:64:dd:b0:00:03 (172.16.1.1): index=4 time=709.368 usec
^C
--- 172.16.1.1 statistics ---
5 packets transmitted, 5 packets received,   0% unanswered (0 extra)
rtt min/avg/max/std-dev = 0.203/0.482/0.844/0.248 ms
[0]-[current]-[/usr/ports/net/arping]
# ldd `which arping`
/usr/local/sbin/arping:
Start            End              Type  Open Ref GrpRef Name
0000126820200000 000012682020c000 exe   1    0   0      /usr/local/sbin/arping                                                                                            0000126b08e33000 0000126b08e6d000 rlib  0    1   0      /usr/lib/libpcap.so.8.4                                                                                           0000126ab9db7000 0000126ab9dd4000 rlib  0    1   0      /usr/local/lib/libnet.so.12.0
0000126a7caec000 0000126a7cb1b000 rlib  0    1   0      /usr/lib/libm.so.10.1
0000126a43d9d000 0000126a43e91000 rlib  0    1   0      /usr/lib/libc.so.95.1
0000126affbbd000 0000126affbbd000 ld.so 0    1   0      /usr/libexec/ld.so
```
Ports file  ( GH_COMMIT give the tested version ):

```
# cat ./Makefile
# $OpenBSD: Makefile,v 1.18 2019/02/08 19:07:21 jasper Exp $

REV=            1.2
VERSION=        ${REV}
SHARED_LIBS=    net     12.0
PKGSPEC=        libnet->=1.2,<1.3v0

BUILD_DEPENDS=  devel/doxygen


GH_ACCOUNT=     libnet
GH_PROJECT=     libnet
GH_COMMIT=      8b85fdac0b4dc8b078cc0c70dcd783fdbead1bdd

WRKDIST=${WRKDIR}/libnet-${GH_COMMIT}

HOME_PAGE=      https://github.com/libnet/libnet

AUTOCONF_VERSION=2.69
AUTOMAKE_VERSION=1.15
AUTOCONF=autoreconf --install

pre-install:
        cd ${WRKSRC} && make doc

MAKE_FLAGS=     LIBnet_VERSION=${LIBnet_VERSION}
CONFIGURE_ENV=  LIBNET_CONFIG_LIBS="-L${LOCALBASE}/lib/libnet-${REV}" \
                LIBNET_CONFIG_CFLAGS="-I${LOCALBASE}/include/libnet-${REV}"

post-install:
        @mkdir -p ${PREFIX}/lib/libnet-1.2/
        @cp ${PREFIX}/lib/libnet.a ${PREFIX}/lib/libnet-1.2/
        @cp ${PREFIX}/lib/libnet.so.${LIBnet_VERSION} ${PREFIX}/lib/libnet-1.2/
        @mkdir -p ${PREFIX}/include/libnet-1.2/libnet
        @mkdir -p ${PREFIX}/include/libnet-1.2/libnet
        @cp ${PREFIX}/include/libnet/* ${PREFIX}/include/libnet-1.2/libnet/
        @cp ${PREFIX}/include/libnet.h ${PREFIX}/include/libnet-1.2/

.include <bsd.port.mk>

```

**Comment by troglobit** [comment]
@dohnuts Hi, as of v1.2 (and current master) we've removed the old `libnet-config` tool in favor of the more standard `pkg-config`.  For each application using libnet they have to do something like this:

    $ pkg-config --libs --static --cflags libnet
    -I/usr/local/include -L/usr/local/lib -lnet

Or, for a GNU autotools based application, use the following in `configure.ac`:

    # Check for required libraries
    PKG_CHECK_MODULES([libnet], [libnet >= 1.2])

and in the `Makefile.am`:

    proggy_CFLAGS = $(libnet_CFLAGS)
    proggy_LDADD  = $(libnet_LIBS)

**Comment by dohnuts** [comment]
That is beyond the scope of libnet.
pkg-config is available on BSD, but package that us libnet will do whatever
theay want:

nothing to report here.

Let's release.

**Comment by troglobit** [comment]
OK.  I've been meaning to get an -rc out but there are still a few things that I'd like to do, namely:

1. `__FreeBSD_version >= 1100030` no longer need byte swap.  See [Porter's Handbook](https://www.freebsd.org/doc/en_US.ISO8859-1/books/porters-handbook/versions-11.html), they've finally fixed the issues with `SOCK_RAW`.  I've yet to investigate DragonflyBSD, NetBSD and OpenBSD
2. Run Valgrind on Nessus, possibly also cross-compile to a [big endian target](https://github.com/troglobit/troglos/tree/master/arch/ppc) to verify STP works
3. I'm a little worried about `libnet-config` tool being missing, for compatibility we should perhaps maintain a wrapper for a time?

**Comment by dohnuts** [comment]
3 . I m not officially a package maintainer , nevertheless the libnet-config seems an overkill, knowing in ports system(i like) it is not required, and apparently in auto lib stuff ( 🙄 ) it s deprecated, so dropping it is the way. Of course this should be discuss with package maintainer.
1/2 I have a small pine64 hardware is sometimes run BSD on it, but building on it would be impossible afaik, maybe a recent raspberry would do the job but i do not have one.

FreeBSD killed two big to little , https://svnweb.freebsd.org/base/head/sys/netinet/raw_ip.c?r1=270008&r2=270929&pathrev=270929 and fixed n internal check

On OpenBSD the behavior looked patch regarding the test :

```
    310         /* Don't allow packet length sizes that will crash. */
    311         if (hlen < sizeof (struct ip) ||
    312             ntohs(ip->ip_len) < hlen ||
    313             ntohs(ip->ip_len) != m->m_pkthdr.len) {
```

Neverhteless, i guess it s firewall related, 

```
    246                 ip->ip_off = htons(0);
    248                 ip->ip_len = htons(m->m_pkthdr.len)
```

https://cvsweb.openbsd.org/cgi-bin/cvsweb/src/sys/netinet/raw_ip.c#rev1.30

the person to contact is itojun.

IMHO/My understanding is that in RAW on OpenBSD all data are untouched endian wise, but the received data len is in host order, for security purpose, if so, it wont be patch as it not acceptable to transfer invalid data.

If you have a test case, that would help ( tcptraceroute, and arping probably do not send silly len size )

**Comment by sam-github** [comment]
1. no comment

2 seems a good idea, but doesn't seem release blocking

3 is hard to know its a problem until release happens

I know I didn't do this due to lack of time, but I'd suggest more of a release often approach if the process can be made fast enough. If there is an outcry about lack of libnet-config, its possible to release again, and if nessus finds some bug that has been lurking unnoticed for years, a patch release to fix it can be done.

**Comment by dohnuts** [comment]
regarding 3 ,  https://github.com/mct/tcptraceroute/blob/master/configure#L3964 , compiling tcptraceroutre will be problematic unless patched.

**Comment by troglobit** [comment]
Good points all around, thank you for the feedback!  I'm not usually not one to discuss things, more of a get-shit-done kinda guy, but I've not been dragging my feet with the release, just want to do my due diligence seeing as this is my first time here.

Also, would be good to hear back from the Windows people about the state of `master`, anyone heard from @sgeto or maybe someone else can test it?  I've set up Appveyor builds based on the work @sgeto did for his tree, as well as automated doc builds and cleaned that up a bit. Any input on all that is also very much welcome.

1. I've spun up one of my virtual FreeBSD machines to test this issue, it's been bothering me in Nemesis, so I'm fixing this one for me.  Awesome to hear about `SOCK_RAW` being fixed and working as the UNIX beards intended also on OpenBSD.  I remember the passing of Itojun way back then, it was really sad.
2. Must've been *really* tired when I wrote this, I meant Nemesis, not Nessus.  Since I'm the maintainer of Nemesis it was really just a quick test with latest master of both projects through Valgrind.  I'll have a look at it tomorrow, but yeah it's optional
3. As of 067fb39 `libnet-config` is back, generated from `libnet-config.in`, now with correct paths to -L and -I

**Comment by dohnuts** [comment]
the libnet-config will make the tcptraceroute diff more easy to read :-)

Nevertheless this requires the libnet port to patch the Makefile to create the 1.2 directory.
( because 1.0 is still in use )

I just use pkg-config and a pc to test tcptraceroute.


```
Index: Makefile
===================================================================
RCS file: /cvs/ports/net/tcptraceroute/Makefile,v
retrieving revision 1.24
diff -u -p -r1.24 Makefile
--- Makefile    12 Jul 2019 20:48:50 -0000      1.24
+++ Makefile    7 Oct 2019 17:57:15 -0000
@@ -11,15 +11,17 @@ HOMEPAGE=           http://michael.toren.net/code
 # GPL
 PERMIT_PACKAGE=        Yes

-WANTLIB=               c pcap lib/libnet-1.1/net=11
+WANTLIB=               c pcap lib/libnet-1.2/net=12

 CONFIGURE_STYLE=       gnu
-CONFIGURE_ARGS=                --with-libnet=${PREFIX} \
-                       --enable-noselect-default
+
+CONFIGURE_ARGS=                --enable-noselect-default \
+                       --with-libnet=${PREFIX}
+

 MASTER_SITES=          http://michael.toren.net/code/tcptraceroute/

-LIB_DEPENDS=           net/libnet/1.1
+LIB_DEPENDS=           net/libnet/1.2

 MAKE_FLAGS=            CC="${CC}" \
                        CFLAGS="${CFLAGS}"
cvs server: Diffing patches
Index: patches/patch-configure
===================================================================
RCS file: /cvs/ports/net/tcptraceroute/patches/patch-configure,v
retrieving revision 1.1
diff -u -p -r1.1 patch-configure
--- patches/patch-configure     26 May 2007 22:06:19 -0000      1.1
+++ patches/patch-configure     7 Oct 2019 17:57:15 -0000
@@ -6,7 +6,7 @@ $OpenBSD: patch-configure,v 1.1 2007/05/
  fi;

 -LIBNET_CONFIG="libnet-config" # relative, using $PATH
-+LIBNET_CONFIG="libnet-config-1.1"     # relative, using $PATH
++LIBNET_CONFIG="pkg-config --libs libnet"      # relative, using $PATH

  # Check whether --with-libnet or --without-libnet was given.
  if test "${with_libnet+set}" = set; then
@@ -14,11 +14,11 @@ $OpenBSD: patch-configure,v 1.1 2007/05/

                test -f "$withval/libnet.h"                     &&      LIBNETCC="$LIBNETCC -I$withval"
                test -f "$withval/include/libnet.h"     &&      LIBNETCC="$LIBNETCC -I$withval/include"
-+              test -f "$withval/include/libnet-1.1/libnet.h"  &&      LIBNETCC="$LIBNETCC -I$withval/include/libnet-1.1"
++              test -f "$withval/include/libnet-1.2/libnet.h"  &&      LIBNETCC="$LIBNETCC -I$withval/include/libnet-1.2"

                test -f "$withval/libnet.a"                     &&      LIBNETLD="$LIBNETLD -L$withval"
                test -f "$withval/lib/libnet.a"         &&      LIBNETLD="$LIBNETLD -L$withval/lib"
-+              test -f "$withval/lib/libnet-1.1/libnet.a"      &&      LIBNETLD="$LIBNETLD -L$withval/lib/libnet-1.1"
++              test -f "$withval/lib/libnet-1.2/libnet.a"      &&      LIBNETLD="$LIBNETLD -L$withval/lib/libnet-1.2"

                if test -z "$LIBNETCC" -o -z "$LIBNETLD"
                then

```

tcptraceroute would nevertheless be better with a real release and little cleaning libtool/autotools wise

**Comment by troglobit** [comment]
@dohnuts Oh dear, what a mess ... oh well, maybe we'll have reduced the pain for some other projects at least.

**Comment by troglobit** [comment]
Status update:

1. Fixed
2. Pending, will look at later today
3. Fixed

So, looks like -rc4 will drop tonight (CET). I'll do some limited testing on that and then we can have a GA release out the door \o/

**Comment by troglobit** [comment]
There, [v1.2-rc4](https://github.com/libnet/libnet/releases/tag/v1.2-rc4) released.  Looking quite good from my side.  Only sticking point is that (html? and) man pages are not built and installed by default if doxygen is found.

Again, someone with any experience with win32 is very welcome to test.

**Comment by sam-github** [comment]
Not sure I understood your first point above.

Isn't the way that it works and should work is that only source files (doc src, code src, autogen src, ...) is comitted to git, autogen.sh and `make distcheck` then create the distributable package that includes the man pages (and html, though whether it really needs to be there is open for debate, I don't think packages need to include html docs, but then, it doesn't really do harm, does it?). So, doxygen is run only by maintainers during release creation, because the releases include the doxygen output.

All of which is to ask, what part of the process should run doxygen if doxygen is found? The "built and installed" is what confuses me, install happens on the side of a release consumer, and should not require doxygen.

**Comment by sam-github** [comment]
And on a slightly related note, I really like how you are pushing releases to https://github.com/libnet/libnet/releases/tag/v1.2-rc4, for example, but I'm not sure how you are doing it. It would be helpful if you could modify doc/RELEASE-CHECKLIST.txt to describe the release process.

I'll dig up the auth credentials for https://sourceforge.net/projects/libnet-dev/ and modify its README to make clear that sourceforge is no longer used and point people at github for releases.

**Comment by troglobit** [comment]
I think we are in 99% agreement.  I don't care much for the html either, other people might like that more but if it's [available online](https://codedocs.xyz/libnet/libnet/) that should be sufficient in my view.  The man pages are more of a sticking point to me and I'm not sure where they should be included.  

Currently I'm leaning towards not including pre-built man pages in GIT (as it is currently), and not even including them in the tarballs.  Only way to get them would be to have doxygen installed and, when doxygen is detected, build man pages (and html) so they are installed with `make install`.  

Because it's not clear to me atm. how to keep generated man pages out of GIT but still be part of the dist tarballs, but I'll read up on it more tonight.  I'm all for making it as easy as possible for an end user that downloads a dist tarball; shell, make and cc should be sufficient to get started.

**Comment by troglobit** [comment]
Thanks, GitHub has neat features for releases and then some not-so-neat features (automatically generated tar/zip archives that have uncommon namespace and not properly generated content).  I was planning to add a doc/HACKING.md file that also details how to step ABI version and stuff.  Something like this --> https://github.com/troglobit/editline/blob/master/docs/HACKING.md

Awesome, updating SourceForge was way down on my list of things to ask you about, great to hear that you're way ahead on that :-)

**Comment by sam-github** [comment]
> how to step ABI version

Very specific information is in the release checklist.

I'm having trouble doing a build from src right now and in do the apt installs in the README, I'm on Ubuntu 19.04, what kind of debian/ubuntu did you use? I'll PR the README, but some words to acknoledge the exact packages may differ by distro will be necessary.

**Comment by troglobit** [comment]
OK, I'll adapt that.  Is libnet-dev-announce still alive?

We might very well have missed something obvious.  I've tested on Ubuntu 19.04, FreeBSD 11 and a couple of embedded targets so far. Wanted to test Debian 10 (and unstable) as well, but haven't gotten around to that yet.  More attention to detail on build deps in the README is *most* welcome :)

**Comment by sam-github** [comment]
> I'm leaning towards not including pre-built man pages in GIT (as it is currently)

I can't find any sign of man pages comitted to git. :question: 

> Because it's not clear to me atm. how to keep generated man pages out of GIT but still be part of the dist tarballs

I think that's actually easy, it just looks hard because doxygen is currently being setup and run by autotools, it doesn't have to be. `./configure` and the rest of autotools output is meant for downstream consumers of a release, releasers can be assumed/required to be running on a modern linux distro.

So, `autogen.sh` could be basically:
```
cd doc && doxgen
autoreconf -W portability -vif $*
```
With the Doxgen conf file in doc/ instead of generated by autotools/configure.

This breaks the cycle. To autotools, the doxgen generated docs will appear to always be there.

My level of interest in wrestling with autotools is very low, and my desire for installed manpages is medium.

Just removing man pages isn't terrible, but a number of packaging conventions request C APIs to have man pages, and I can see the desire for that, I like good docs! Not to imply that libnet's are :-).

For example, `man doc/man/man3/libnet.3` refers to the manpage for `libnet-config`, but that doesn't exist, it probably needs updating to refer to pkg-config

**Comment by troglobit** [comment]
Yeah, sorry my English is not always up to par, in particular after a whole day at $DAYJOB. The man pages, like the html, has been removed from GIT as of our previous discussion on that (a week ago or so).

I can wrestle autotools, no problem.  I just want it to be readable by others as well :)  My goal is to make everything as simple as possible, even for a maintainer.  Which to me means; if doxygen is installed and user types `make release` everything magic should happen: 1) `make doc`, 2) `make distcheck`, and 3) check for git tag.  Possibly even propose ot push.

Thanks for the heads-up on the reference to libnet-config!

**Comment by sam-github** [comment]
sounds reasonable. I'm having build problems, see #90 

Also, I did `make distcheck`, then tried to build the generated tarball, and `./configure` said at the end to call `make doc`, but it didn't work:
```
__/libnet-1.2-rc4 % make doc
  DXGEN  Doxyfile
error: tag HTML_HEADER: header file `./doc/libnet_api_header.html_tmpl' does not exist
make: *** [Makefile:1008: doc/libnet.tag] Error 1
```

Did I generate the tgz wrongly? That checklist needs updating :-)

**Comment by troglobit** [comment]
Uh, that's no good ... looks like distclean might have removed the template file ... :-/

**Comment by troglobit** [comment]
Fixed in 536d6c8, was missing from EXTRA_DIST.  I'll continue looking into automating/streamlining man pages in dist, if possible.

**Comment by sam-github** [comment]
I updated https://sourceforge.net/p/libnet-dev/admin/ to make clear that github is the new home.

I tried to post to libnet-announce to say that the sourceforge is dead, but can't get into it, and the admin access page is broken on sourceforge. So it goes.

**Comment by sam-github** [comment]
> Release curated by Sam Roberts and Joachim Nilsson.

Btw, I've no objections to the above, but if you want to simplify and automate the release process and not hand-write thank you notes, that's OK with me. My identity in the git history is all the credit (embaressment?) I need.

**Comment by troglobit** [comment]
Cool, https://sourceforge.net/projects/libnet-dev/ looks great now, thanks for taking the time to fix this :)

Btw, you're too modest. You've overseen the project and the 1.2 release this far, for the last decade. You definitely deserve an honorable mention in the release notes.  I haven't really decided yet on automation vs hand-writing it, may be a little bit of both since I strongly believe in having useful/readable release notes that are not just `git log >> release-notes`.

I've fixed the issue with building and distributing the man pages in 8dc3678.  I'll have a look at the docs now, see if there's anything else (beside what's been mentioned previously here) I can fix before the GA.  Otherwise we're rapidly approaching the GA release! \o/

**Comment by troglobit** [comment]
@sam-github OK, after merging your PR #92, I've now completed the magic for doc + distcheck on master as of f406201. Travis seems happy (although the macOS build takes a bit long due to the `brew update`. It's unfortunately quite a bit late here in Sweden now, so I'll get back to the reminder of the release work tomorrow evening after $DAYJOB, but it looks really good for GA release this weekend :smiley::+1:

**Comment by sam-github** [comment]
`git clean -x -d && sh autogen.sh && ./configure && make release` still doesn't work:

> No rule to make target 'libnet.3', needed by 'distdir-am'.  Stop.

But `git clean -x -d && sh autogen.sh && ./configure && make doc release` works swimmingly... perhaps the release target should depend on the doc target, but that's just niceties, its building fine now.

**Comment by troglobit** [comment]
Get into the same problem with `make distclean; ./configure && make distcheck`.  Think I've just fixed it ... as well as given the macOS builds a bit of a speedup :)

**Comment by troglobit** [comment]
There, final touches updates/renames a few doc files.  Looks like we're finally ready for release!  I'm off to work and will do the GA release later tonight, CET, unless I hear any objections or reminder of stuff discussed here that I've forgot :smiley:

**Comment by sam-github** [comment]
Expected?

1. `git clean -x -d&&sh autogen.sh&&./configure&&make release`
2. `cd ..`
3. `rm -rf libnet-1.2-rc4 && tar -xf libnet-1.2-rc4.tar.gz&& cd libnet-1.2-rc4 && ./configure && make install DESTDIR=/tmp/libnet-1.2-rc4`

```
Making install in man1
make[4]: Nothing to be done for 'install-exec-am'.
 /bin/mkdir -p '/tmp/libnet-1.2-rc4/usr/local/share/man/man3'
 /usr/bin/install -c -m 644 'libnet-config.1' '/tmp/libnet-1.2-rc4/usr/local/share/man/man3/libnet-config.3'                                                                 
Making install in man3
make[4]: Nothing to be done for 'install-exec-am'.
 /bin/mkdir -p '/tmp/libnet-1.2-rc4/usr/local/share/man/man3'
 /usr/bin/install -c -m 644 ./libnet.3 ./libnet-functions.3 ./libnet-macros.3 '/tmp/libnet-1.2-rc4/usr/local/share/man/man3'                                                 
make[4]: Nothing to be done for 'install-exec-am'.
make[4]: Nothing to be done for 'install-data-am'.
make[2]: *** No rule to make target 'html/*', needed by 'all-am'.  Stop.
make[1]: *** [Makefile:467: install-recursive] Error 1
```

I tried `make doc  && make release` with the same result.

How do you build a release?

**Comment by sam-github** [comment]
On the side of downstream users of the package .tgz, they have to do `make doc install`.

I think this is because `make install` is currently trying to install both man pages (which are included in the .tgz) and html (which is not included in the tgz), thus the
```
make[2]: *** No rule to make target 'html/*', needed by 'all-am'.  Stop.
```

**Comment by troglobit** [comment]
Good point, and thank you for testing!  I had postponed testing that part of the release process and it's clearly not working right yet.  I'll have a look at it right away, because the flow you describe is what everyone expect to just work.

**Comment by troglobit** [comment]
Should be fixed in b796e3b, now built and installed from both GIT and dist tarball if enabled and doxygen is available. Hopefully I didn't miss any obvious step ... :pray:

**Comment by sam-github** [comment]
Built and installed, looking great.

I did notice that the html docs went in `/usr/local/share/libnet/...`, right beside `/usr/local/share/man/...`, is that right? I would expect `/usr/local/share/doc/libnet/html/...`, which would be essentially what debian does, not a new top-level directory in `/usr/local/share`.

Just a suggestoin, distros will use the configure scripts to put these things exactly where they want.

**Comment by sam-github** [comment]
FYI, useful?

```
p/libnet (master *% u=) % git diff
diff --git a/bin/check-release b/bin/check-release
new file mode 100755
index 0000000..0cdc790
--- /dev/null
+++ b/bin/check-release
@@ -0,0 +1,14 @@
+#!/bin/sh
+
+set -e
+set -x
+
+git clean -x -d
+sh autogen.sh
+./configure
+make release
+rm -rf libnet-1.2-rc4
+tar -xf ../libnet-1.2-rc4.tar.gz
+cd libnet-1.2-rc4
+./configure
+make install DESTDIR=/tmp/libnet-1.2-rc4
```

**Comment by troglobit** [comment]
You have a keen eye!  Must've been some cut-n-paste error, fixed install path for HTML to default to `/usr/local/share/doc/libnet/html/` (in your example). 

I was thinking of something similar to your `check-release` script, but I'd really like to have it; a) in Travis-CI or maybe in the release target, and b) be `$(VERSION)` independent.  Was actually playing around with it an hour ago.  I'll see what I can cook up :-)

**Comment by troglobit** [comment]
Regarding your script, I think `make distcheck` does this already, and distcheck is a dependency for the release target, which is run by Travis. Looking atm at why we didn't catch your issue there

**Comment by troglobit** [comment]
Don't understand why we didn't catch your issue in Travis, the latest build log looks great:

   https://travis-ci.org/libnet/libnet/jobs/598327733

It runs both `distcheck` and `installcheck`, which builds doc and installs to a separate directory fine.
I say we ship it.

**Comment by troglobit** [comment]
There, released!  \o/

Huge thanks to everyone who's been involved, contributed patches, tested and hung in there in general!
